File: MIDILink/MIDISystem.m
//
//  MIDISystem.m
//  MIDILink
//
//  Created by Julian Porter on 26/04/2016.
//  Copyright © 2016 JP Embedded Solutions. All rights reserved.
//

#import "MIDISystem.h"
#import "MIDIThing.h"
#import "Endpoint.h"


@implementation MIDISystem

+ (void) errorWithCode:(OSStatus)code andDescription:(NSString *)description {
    [[NSException exceptionWithName:@"MIDI Error"
                             reason:[NSString stringWithFormat:@"Error %d - %@",code,description]
                           userInfo:nil ] raise];
}




+ (NSArray *)enumerate:(MIDIDeviceType)type {
    ItemCount n;
    NSMutableArray *items=[NSMutableArray array];
    NSLog(@"Enumerating . . .");
    switch (type) {
        case MIDIDeviceTypeSource: {
            n=MIDIGetNumberOfSources();
            for(ItemCount i=0;i<n;i++) {
                MIDIEndpointRef device=MIDIGetSource(i);
                MIDIThing *thing=[MIDIThing fromObject:device];
                [items addObject:[[MIDIEndPointDescription alloc] initWithThing:thing]];
                
            }
            break; }
        case MIDIDeviceTypeDestination: {
            n=MIDIGetNumberOfDestinations();
            for(ItemCount i=0;i<n;i++) {
                MIDIEndpointRef device=MIDIGetDestination(i);
                MIDIThing *thing=[MIDIThing fromObject:device];
                [items addObject:[[MIDIEndPointDescription alloc] initWithThing:thing]];

            }
            break; }
        case MIDIDeviceTypeAll: {
            n=MIDIGetNumberOfSources();
            NSLog(@"NSources = %ld",n);
            for(ItemCount i=0;i<n;i++) {
                MIDIEndpointRef device=MIDIGetSource(i);
                MIDIThing *thing=[MIDIThing fromObject:device];
                [items addObject:[[MIDIEndPointDescription alloc] initWithThing:thing]];
                
            }
            n=MIDIGetNumberOfDestinations();
            NSLog(@"NDestinations = %ld",n);
            for(ItemCount i=0;i<n;i++) {
                MIDIEndpointRef device=MIDIGetDestination(i);
                MIDIThing *thing=[MIDIThing fromObject:device];
                [items addObject:[[MIDIEndPointDescription alloc] initWithThing:thing]];
            }
            break; }
    }
    NSLog(@"Item has length %ld",[items count]);
    for(MIDIEndPointDescription *d in items) NSLog(@"%@",d);
    return items;
    
}

@end

File: MIDITools/System/MIDISystem.swift
//
//  MIDISystem.swift
//  MIDI Utils
//
//  Created by Julian Porter on 14/02/2017.
//  Copyright © 2017 JP Embedded Solutions. All rights reserved.
//

import Foundation
import CoreMIDI



public class MIDISystem {
    private static var the : MIDISystem?
    
    public private(set) var _sources : [MIDIEndpoint] = []
    public private(set) var _destinations : [MIDIEndpoint] = []
    
    public init() throws {
        var n=MIDIGetNumberOfSources();
        for i in 0..<n  {
            let device=MIDIGetSource(i)
            if let object  = try getMIDIObject(device) as? MIDIEndpoint { _sources.append(object) }
        }
        n=MIDIGetNumberOfDestinations();
        for i in 0..<n  {
            let device=MIDIGetDestination(i)
            if let object = try getMIDIObject(device) as? MIDIEndpoint { _destinations.append(object) }
        }
    }
    public static var sources : [MIDIEndpoint] { the?._sources ?? [] }
    public static var destinations : [MIDIEndpoint] { the?._destinations ?? [] }
    public static var endpoints : Set<MIDIEndpoint> { Set(sources).union(destinations) }
    public static func scan() throws { the=try MIDISystem() }
}


